### PR TITLE
Do not save gear favorites

### PR DESCRIPTION
**Pull request by CaffeineBytes**
Prevents saving to favorites files for players and bots. Keep bots favorites saved in-game until quit.

### CODE REVIEW

File: GTFO-API/Patches/PlayerBackpackManager_Patches.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        [HarmonyWrapSafe]
+        [HarmonyPrefix]
+        public static bool GiveFavoriteGearToBotPrefix(SNet_Player bot, ref bool __result)
+        {
+            var botIndex = bot.PlayerSlotIndex();
+
+            if (string.IsNullOrEmpty(GearManager_Patches.BotFavorites[InventorySlot.GearMelee][botIndex]))
+            {
+                PlayerBackpackManager.CopyGearToBot(bot, PlayerBackpackManager.LocalBackpack);
+            }
```
Comment (Kasuromi) [inline GTFO-API/Patches/PlayerBackpackManager_Patches.cs:20]:
```suggestion
            }
```

File: GTFO-API/Patches/PlayerBackpackManager_Patches.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+{
+    [HarmonyPatch(typeof(PlayerBackpackManager))]
+    internal class PlayerBackpackManager_Patches
+    {
+        [HarmonyPatch(nameof(PlayerBackpackManager.GiveFavoriteGearToBot))]
+        [HarmonyWrapSafe]
+        [HarmonyPrefix]
+        public static bool GiveFavoriteGearToBotPrefix(SNet_Player bot, ref bool __result)
+        {
+            var botIndex = bot.PlayerSlotIndex();
```
Comment (Kasuromi) [inline GTFO-API/Patches/PlayerBackpackManager_Patches.cs:15]: Explicitly declare the field type
```suggestion
            int botIndex = bot.PlayerSlotIndex();
```

File: GTFO-API/Patches/GearManager_Patches.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+using HarmonyLib;
+using Player;
+using UnhollowerBaseLib.Attributes;
+
+namespace GTFO.API.Patches
+{
+    [HarmonyPatch(typeof(GearManager))]
+    internal class GearManager_Patches
+    {
+        [HideFromIl2Cpp]
```
Comment (Kasuromi) [inline GTFO-API/Patches/GearManager_Patches.cs:12]: HideFromIl2Cpp attribute isn't necessary, the patch doesn't get injected into Il2Cpp domain

File: GTFO-API/Patches/GearManager_Patches.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            BotFavorites = new Dictionary<InventorySlot, string[]>
+            {
+                { InventorySlot.GearMelee, new string[4] },
+                { InventorySlot.GearStandard, new string[4] },
+                { InventorySlot.GearSpecial, new string[4] },
+                { InventorySlot.GearClass, new string[4] },
+                { InventorySlot.HackingTool, new string[4] }
+            };
+
+            GearManager.FavoritesData = new GearFavoritesData();
```
Comment (Kasuromi) [inline GTFO-API/Patches/GearManager_Patches.cs:29]: Omit type names and initialize the dictionary by key to value
```suggestion
            BotFavorites = new()
            {
                [InventorySlot.GearMelee] = new string[4],
                [InventorySlot.GearStandard] = new string[4],
                [InventorySlot.GearSpecial] = new string[4],
                [InventorySlot.GearClass] = new string[4],
                [InventorySlot.HackingTool] = new string[4]
            };

            GearManager.FavoritesData = new();
```

File: GTFO-API/Patches/PlayerBackpackManager_Patches.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace GTFO.API.Patches
+{
+    [HarmonyPatch(typeof(PlayerBackpackManager))]
+    internal class PlayerBackpackManager_Patches
+    {
+        [HarmonyPatch(nameof(PlayerBackpackManager.GiveFavoriteGearToBot))]
+        [HarmonyWrapSafe]
+        [HarmonyPrefix]
+        public static bool GiveFavoriteGearToBotPrefix(SNet_Player bot, ref bool __result)
```
Comment (Kasuromi) [inline GTFO-API/Patches/PlayerBackpackManager_Patches.cs:13]: Follow the patch naming convention
```suggestion
        public static bool GiveFavoriteGearToBot_Prefix(SNet_Player bot, ref bool __result)
```

File: GTFO-API/Patches/GearManager_Patches.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+using Player;
+using UnhollowerBaseLib.Attributes;
+
+namespace GTFO.API.Patches
+{
+    [HarmonyPatch(typeof(GearManager))]
+    internal class GearManager_Patches
+    {
+        [HideFromIl2Cpp]
+        public static Dictionary<InventorySlot, string[]> BotFavorites { get; set; }
```
Comment (Kasuromi) [inline GTFO-API/Patches/GearManager_Patches.cs:13]: I'm not sure how I feel about storing persistent data like this in a patch.
My thought is that we should have a PersistentData static class in the Resources namespace where we store this.
I can merge this without changes and then reimplement it a different time

**Review by Kasuromi — CHANGES_REQUESTED** [review]
Happy to merge once changes are applied

File: GTFO-API/Patches/GearManager_Patches.cs (hunk 6 above)
Comment (CaffeineBytes) [reply GTFO-API/Patches/GearManager_Patches.cs]: I like it, maybe in the future there would be a storage api that could use this more. Created and moved bot favorites there.

File: GTFO-API/Resources/PersistentData.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿using System.Collections.Generic;
+using Player;
+
+namespace GTFO.API.Resources
+{
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
+    public static class PersistentData
```
[the remark this answers is not in the document]
Comment (CaffeineBytes) [reply GTFO-API/Resources/PersistentData.cs]: Resolved in [681a1eb](https://github.com/Kasuromi/GTFO-API/pull/27/commits/681a1ebae586bbe5221fd64bb71fbb4ebd56e1bf)

**Pull request merged**